objects/src/accounts/storage/mod.rs

```diff
@@ -281,11 +290,73 @@ impl AccountStorage {
             return Err(AccountError::StorageSlotIsReserved(index));
         }
 
+        // only value slots of basic arity can currently be updated
+        match self.layout[index as usize] {
+            StorageSlotType::Value { value_arity } => {
+                if value_arity > 0 {
+                    return Err(AccountError::StorageSlotInvalidValueArity {
+                        slot: index,
+                        expected: 0,
+                        actual: value_arity,
+                    });
+                }
+            },
+            StorageSlotType::Map { value_arity } => {
+                if value_arity > 0 {
+                    return Err(AccountError::StorageSlotInvalidValueArity {
+                        slot: index,
+                        expected: 0,
+                        actual: value_arity,
+                    });
+                }
+            },
```

Comment:
I don't think we should allow setting maps directly. All updates could go through `set_map_item()` or `apply_delta()`.